### PR TITLE
Add labels.IsDNS1123Label return value  (bool, error), for easy to le…

### PR DESCRIPTION
**Pull request by edwardpro**
…arn invalid reason.

### CODE REVIEW

File: pkg/config/labels/instance.go [hunk 1]
```diff
@@ -92,14 +92,20 @@ func (i Instance) Validate() error {
 
 // IsDNS1123Label tests for a string that conforms to the definition of a label in
 // DNS (RFC 1123).
-func IsDNS1123Label(value string) bool {
-	return len(value) <= DNS1123LabelMaxLength && dns1123LabelRegexp.MatchString(value)
+func IsDNS1123Label(value string) (bool, error) {
+	if len(value) <= DNS1123LabelMaxLength && dns1123LabelRegexp.MatchString(value) {
+		return true, nil
+	}
+	return false, fmt.Errorf("dns label length is > %d by rfc 1123 or had invalid character", DNS1123LabelMaxLength)
```
Comment (esnible) [inline pkg/config/labels/instance.go]: Capitalize DNS.  Istio's lint will complain, so add `// nolint: golint,stylecheck` to the line.

File: pkg/config/labels/instance.go [hunk 2]
```diff
@@ -92,14 +92,20 @@ func (i Instance) Validate() error {
 
 // IsDNS1123Label tests for a string that conforms to the definition of a label in
 // DNS (RFC 1123).
-func IsDNS1123Label(value string) bool {
-	return len(value) <= DNS1123LabelMaxLength && dns1123LabelRegexp.MatchString(value)
+func IsDNS1123Label(value string) (bool, error) {
+	if len(value) <= DNS1123LabelMaxLength && dns1123LabelRegexp.MatchString(value) {
```
Comment (esnible) [inline pkg/config/labels/instance.go]: The error message(s) would be more clear if you break this in two -- check for DNS1123LabelMaxLength and report if there is a problem, check for MatchString and report that problem separately.

File: pkg/config/labels/instance.go [hunk 3]
```diff
@@ -92,14 +92,20 @@ func (i Instance) Validate() error {
...
 
 // IsWildcardDNS1123Label tests for a string that conforms to the definition of a label in DNS (RFC 1123), but allows
 // the wildcard label (`*`), and typical labels with a leading astrisk instead of alphabetic character (e.g. "*-foo")
-func IsWildcardDNS1123Label(value string) bool {
-	return len(value) <= DNS1123LabelMaxLength && wildcardPrefixRegexp.MatchString(value)
+func IsWildcardDNS1123Label(value string) (bool, error) {
+	if len(value) <= DNS1123LabelMaxLength && wildcardPrefixRegexp.MatchString(value) {
+		return true, nil
+	}
+	return false, fmt.Errorf("wildcard prefix dns label length is > %d by rfc 1123 or had invalid character", DNS1123LabelMaxLength)
```
Comment (esnible) [inline pkg/config/labels/instance.go]: Same comments as above.

File: pilot/pkg/model/validation.go [hunk 4]
```diff
@@ -38,8 +38,8 @@ func (s *Service) Validate() error {
 	}
 	parts := strings.Split(string(s.Hostname), ".")
 	for _, part := range parts {
-		if !labels.IsDNS1123Label(part) {
-			errs = multierror.Append(errs, fmt.Errorf("invalid hostname part: %q", part))
+		if ok, err := labels.IsDNS1123Label(part); !ok {
+			errs = multierror.Append(errs, fmt.Errorf("invalid hostname part: %q, reason: %v", part, err))
```
Comment (esnible) [inline pilot/pkg/model/validation.go]: I don't object to "reason:", but typically in Go the error prefixes the underlying error.  Consider `invalid hostname part %q: %v` instead.  Consider this for all of your messages.

**Comment by esnible** [comment]
Thanks for this PR, @edwardpro.  For legal reasons we need to get @zeyin involved with the PR as well.  Can you set this up?

File: pkg/config/labels/instance.go [hunk 5]
```diff
@@ -92,14 +92,20 @@ func (i Instance) Validate() error {
 
 // IsDNS1123Label tests for a string that conforms to the definition of a label in
 // DNS (RFC 1123).
-func IsDNS1123Label(value string) bool {
-	return len(value) <= DNS1123LabelMaxLength && dns1123LabelRegexp.MatchString(value)
+func IsDNS1123Label(value string) (bool, error) {
```
Comment (nmittler) [inline pkg/config/labels/instance.go]: This is awkward. If you need to return an error, consider just changing this to a `Check` style method:

```golang
func CheckDNS1123Label(value string) error {
...
}
```

Call sites that are just looking for true/false will then just check the error against `nil`.

**Pull request closed without merging**